compiler/rustc_hir_typeck/src/fallback.rs

```diff
@@ -464,6 +472,46 @@ impl<'tcx> FnCtxt<'_, 'tcx> {
         diverging_fallback
     }
 
+    fn lint_obligations_broken_by_never_type_fallback_change(
+        &self,
+        behavior: DivergingFallbackBehavior,
+        diverging_vids: &[TyVid],
+    ) {
+        let DivergingFallbackBehavior::FallbackToUnit = behavior else { return };
+
+        // Returns errors which happen if fallback is set to `fallback`
+        let try_out = |fallback| {
+            self.probe(|_| {
+                let obligations = self.fulfillment_cx.borrow().pending_obligations();
+                let ocx = ObligationCtxt::new(&self.infcx);
+                ocx.register_obligations(obligations.iter().cloned());
+
+                for &diverging_vid in diverging_vids {
+                    let diverging_ty = Ty::new_var(self.tcx, diverging_vid);
+
+                    _ = ocx.eq(&ObligationCause::dummy(), self.param_env, diverging_ty, fallback);
+                }
+
+                ocx.select_where_possible()
+            })
+        };
+
+        // If we have no errors with `fallback = ()`, but *do* have errors with `fallback = !`,
+        // then this code will be broken by the never type fallback change.qba
+        let unit_errors = try_out(self.tcx.types.unit);
```

Comment:
So, I have to go back and look at this code path, but this is likely going to be way too slow perf.

First, we're going to end up duplicating the obligation processing, which alone is problematic.

Second, we likely only want to try to lint about fallback if there *was* any fallback (if there is no fallback, then you would expect there to be no change in errors if you change the fallback type).

Reply:
> First, we're going to end up duplicating the obligation processing, which alone is problematic.

It really depends on how many obligations remain stalled after fallback. I would expect there are not many. I personally wouldn't particularly worry about perf here.

Reply:
I could add a fast-exit for cases where `diverging_vids.is_empty()` (i.e. the only case where fallback won't happen), but as compiler-errors said, I don't think this will be perf sensitive.

Reply:
A fast exit will actually be beneficial regardless if it's perf-sensitive. I'd rather you add that since it's not too difficult.

Reply:
fast exit added 👍🏻